flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java

```diff
@@ -0,0 +1,328 @@
+                    try {
+                        return TestingExecutionGraphBuilder.newBuilder().build();
+                    } catch (Throwable e) {
+                        throw new RuntimeException("Error", e);
+                    }
+                };
+        private final List<Tuple3<State, Runnable, Duration>> actions = new ArrayList<>();
+
+        private void transitionToState(State target) {
+            if (transitionTo != null) {
+                Assert.fail("More than one state transition initiated");
+            }
+            transitionTo = target;
+            waitingForResources.onLeave(target);
+        }
+
+        private State getTransitionTo() {
+            return transitionTo;
+        }
+
+        @Override
+        public void goToFinished(ArchivedExecutionGraph archivedExecutionGraph) {
+            transitionToState(new Finished(null, archivedExecutionGraph, LOG));
+        }
+
+        @Override
+        public void goToExecuting(ExecutionGraph executionGraph) {
+            Executing.Context mockExecutingContext = new MockExecutingContext();
+            transitionToState(
+                    new Executing(
+                            executionGraph,
+                            new ExecutionGraphHandler(
+                                    executionGraph, LOG, ForkJoinPool.commonPool()),
+                            new OperatorCoordinatorHandler(
+                                    executionGraph,
+                                    (throwable -> {
+                                        throw new RuntimeException("Error in test", throwable);
+                                    })),
+                            LOG,
+                            mockExecutingContext,
+                            ClassLoader.getSystemClassLoader()));
+        }
+
+        @Override
+        public ArchivedExecutionGraph getArchivedExecutionGraph(
+                JobStatus jobStatus, @Nullable Throwable cause) {
+            return new ArchivedExecutionGraphBuilder()
+                    .setState(jobStatus)
+                    .setFailureCause(cause == null ? null : new ErrorInfo(cause, 1337))
+                    .build();
+        }
+
+        @Override
+        public boolean hasEnoughResources(ResourceCounter desiredResources) {
+            return hasEnoughResourcesSupplier.get();
+        }
+
+        @Override
+        public ExecutionGraph createExecutionGraphWithAvailableResources() throws Exception {
+            return createExecutionGraphWithAvailableResources.get();
+        }
+
+        @Override
+        public void runIfState(State expectedState, Runnable action, Duration delay) {
+            actions.add(Tuple3.of(expectedState, action, delay));
+        }
+
+        // ---- Testing extensions ------
+
+        public void setWaitingForResources(WaitingForResources wfr) {
+            this.waitingForResources = wfr;
+        }
+
+        public List<Tuple3<State, Runnable, Duration>> getActions() {
```

Comment:
I guess you are using tuples because it is a wip branch. In the final PR, please don't use tuples. They make reading code unnecessarily hard.